lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90VectorReader.java

```diff
@@ -0,0 +1,326 @@
+    return new OffHeapVectorValues(fieldEntry, bytesSlice);
+  }
+
+  @Override
+  public void close() throws IOException {
+    vectorData.close();
+  }
+
+  private static class FieldEntry {
+
+    final int dimension;
+    final VectorValues.ScoreFunction scoreFunction;
+    final int maxDoc;
+
+    final long vectorDataOffset;
+    final long vectorDataLength;
+    final int[] ordToDoc;
+
+    FieldEntry(int dimension, VectorValues.ScoreFunction scoreFunction, int maxDoc,
+               long vectorDataOffset, long vectorDataLength, int[] ordToDoc) {
+      this.dimension = dimension;
+      this.scoreFunction = scoreFunction;
+      this.maxDoc = maxDoc;
+      this.vectorDataOffset = vectorDataOffset;
+      this.vectorDataLength = vectorDataLength;
+      this.ordToDoc = ordToDoc;
+    }
+
+    int size() {
+      return ordToDoc.length;
+    }
+  }
+
+  /** Read the vector values from the index input. This supports both iterated and random access. */
+  private final static class OffHeapVectorValues extends VectorValues {
+
+    final FieldEntry fieldEntry;
+    final IndexInput dataIn;
+
+    final BytesRef binaryValue;
+    final ByteBuffer byteBuffer;
+    final FloatBuffer floatBuffer;
+    final int byteSize;
+    final float[] value;
+
+    int ord = -1;
+    int doc = -1;
+
+    OffHeapVectorValues(FieldEntry fieldEntry, IndexInput dataIn) {
+      this.fieldEntry = fieldEntry;
+      this.dataIn = dataIn;
+      byteSize = Float.BYTES * fieldEntry.dimension;
+      byteBuffer = ByteBuffer.allocate(byteSize);
+      floatBuffer = byteBuffer.asFloatBuffer();
+      value = new float[fieldEntry.dimension];
+      binaryValue = new BytesRef(byteBuffer.array(), byteBuffer.arrayOffset(), byteSize);
+    }
+
+    @Override
+    public int dimension() {
+      return fieldEntry.dimension;
+    }
+
+    @Override
+    public int size() {
+      return fieldEntry.size();
+    }
+
+    @Override
+    public ScoreFunction scoreFunction() {
+      return fieldEntry.scoreFunction;
+    }
+
+    @Override
+    public float[] vectorValue() throws IOException {
+      dataIn.seek(ord * byteSize);
+      dataIn.readBytes(byteBuffer.array(), byteBuffer.arrayOffset(), byteSize);
+      floatBuffer.position(0);
+      floatBuffer.get(value, 0, fieldEntry.dimension);
+      return value;
+    }
+
+    @Override
+    public BytesRef binaryValue() throws IOException {
+      dataIn.seek(ord * byteSize);
+      dataIn.readBytes(byteBuffer.array(), byteBuffer.arrayOffset(), byteSize);
+      return binaryValue;
+    }
+
+    @Override
+    public int docID() {
+      return doc;
+    }
+
+    @Override
+    public int nextDoc() {
+      if (++ord >= size()) {
+        doc = NO_MORE_DOCS;
+      } else {
+        doc = fieldEntry.ordToDoc[ord];
+      }
+      return doc;
+    }
+
+    @Override
+    public int advance(int target) throws IOException {
+      // We could do better by log-binary search in ordToDoc, but this is never used
```

Comment:
Yes, the doc-> ord relation is monotonic, ie order-preserving s.t. `doc1 > doc2 iff ord(doc1) > ord(doc2)`. We use this fact in the implementation of random-access over a sorted index while merging so I don't think we're free to share vectors like that (with this implementation). I also plan to make use of that invariant when building NSW graphs - we want to store the ords in sorted order, and if we generate the graph while indexing, then they will naturally be added in order (since their order corresponds to the docid order) with no need to sort them at any point.

I tend to think it's unlikely that you would often see duplicated floating-point vectors aside from zeros, although of course YMMV. I think we will want to allow for lossy compression using byte-per-dimension rather than float-per-dimension, but I am hard-pressed to see what kind of lossless compression we could provide. Well I see there is stuff out there: https://computing.llnl.gov/projects/floating-point-compression , but I expect it would require some additional index structure to enable random access